Fixed bugs and added jsoup

PaaaulZ opened the pull request:
- Fixed crash on retrieve by link (streamscharts/twitchtracker). Fixes: #193
- Removed references to /v5/ api soon to be deprecated. Fixes: #194 


- Added jsoup to ease the information gathering from  streamscharts/twitchtracker

mgonch commented:
Would you mind compiling this version into an exe?

sam1677 commented on src/TwitchRecover.Core/WebsiteRetrieval.java:
```diff
@@ -172,40 +129,70 @@ private static String getJSON(String url) throws IOException {
-}
\ No newline at end of file
+
+    private static Integer airTimeFromString(String airTimeString) {
```
I think it is better to use multiple groups.
`([0-9]+h)([0-9]+m)`

sam1677 replied:
ah streams chart trims (\d+h) if the hour is zero

PaaaulZ replied:
You are right, that looks cleaner. Resolved with commit [bb242e7](https://github.com/TwitchRecover/TwitchRecover/pull/195/commits/bb242e793b73720fb0921393e9e0d2712533c33b)

I changed your regex a bit because it had problems with a stream with less than 1h airtime

PaaaulZ commented:
> Would you mind compiling this version into an exe?

You can find the exe in my releases section [HERE](https://github.com/PaaaulZ/TwitchRecover/releases/download/2.0aH/TwitchRecover.exe)

The .exe and .jar already includes fixes suggested by @sam1677

mgonch commented:
Hmm, it's still crashing on my end with that release.

PaaaulZ commented:
Can you send me the url you tried? I tried both streamcharts and twitchtracker this morning and worked

mgonch commented:
https://twitchtracker.com/kswinniie/streams/44769867132

PaaaulZ commented:
You're right, it worked with option 15 but with option 5 crashed because it needs a startdate that I forgot to add for TwitchTracker. Fixed with this commit, you can find the updated .exe [in the release section](https://github.com/PaaaulZ/TwitchRecover/releases/download/2.0aH/TwitchRecover.exe)

When you are done testing and ready to merge let me know if you want me to squash all these commits into one

KaMyKaSii commented:
@PaaaulZ I tried your fork with a [streamscharts](https://streamscharts.com/channels/trizpariz/streams/45480728909) link and it didn't work. A shell script I wrote works for the same link and as you can see it is playable. Could you see what the problem is? Thanks
![Screenshot from 2022-03-08 17-07-06](https://user-images.githubusercontent.com/17092335/157316960-ad727fb3-4402-47be-a772-54b5c48ad285.png)
![Screenshot from 2022-03-08 17-07-35](https://user-images.githubusercontent.com/17092335/157316975-bacefba0-3230-401e-979f-d6d8eb9f936b.png)
[get-vod.zip](https://github.com/TwitchRecover/TwitchRecover/files/8209294/get-vod.zip)

KaMyKaSii commented:
@PaaaulZ can you make a new release with the fix? Thanks

PaaaulZ commented:
> @PaaaulZ can you make a new release with the fix? Thanks

I updated the old one on my repository

KaMyKaSii commented:
> I updated the old one on my repository

Thanks. As you seem to want to improve TwitchRecover, could you activate Issues in your fork? One feature I would like to ask is that since Twitch sometimes keeps the index-dvr.m3u8 file but deletes the ts files, that you add an option to test if the .ts files are available (perhaps along with the unmute option) and then produce an .m3u8 file without the unavailable .ts files, or return an error message if all .ts files have been deleted.

PaaaulZ commented:
Since this project is not officially abandoned I would like to stick to this one and make pull requests, I would like to try keeping this one alive. If you don't mind open the issue here with an example and when I get a chance I'll try to implement it and make a pr :)

If one day they declare the project abandoned maybe I'll stick to my fork

Drimon99 commented:
is the jar file in the releases section updated? because I still get the following error with streamscharts link:
Exception in thread "main" java.lang.NullPointerException
        at TwitchRecover.Core.WebsiteRetrieval.getSCData(WebsiteRetrieval.java:166)
        at TwitchRecover.Core.WebsiteRetrieval.getData(WebsiteRetrieval.java:64)
        at TwitchRecover.Core.VOD.retrieveVODURL(VOD.java:141)
        at TwitchRecover.CLI.Handlers.VODHandler.recover(VODHandler.java:96)
        at TwitchRecover.CLI.Handlers.VODHandler.<init>(VODHandler.java:44)
        at TwitchRecover.CLI.CLIHandler.main(CLIHandler.java:68)
        at TwitchRecover.CLI.CLI.main(CLI.java:30)

PaaaulZ commented:
@Drimon99 the one in the release section of [my repository](https://github.com/PaaaulZ/TwitchRecover/releases/tag/2.0aH) is updated, this one is not.
If mine does not work send the link here and I'll check

Drimon99 commented:
yes it was that, but I was wrong, the error came out when I tried to insert the values manually with channel name, stream id and start time. if I use the streamschart link this error comes out:
NO RESULTS FOUND.
The VOD cannot be found on Twitch servers.

PaaaulZ commented:
It happens, sometimes the vod gets deleted permanently on Twitch servers

Drimon99 commented:
I think it was deleted because the streamer was drunk and maybe he didn't want to keep the vod, but the live wasn't over yet. However your new version works with streamscharts links, the only "error" I found is that if you use the links to recover a vod and it isn't found "NO RESULTS FOUND" is displayed, but if you enter it manually it closes without displaying errors.

johnc6209 commented:
It's still crashing on fixed release.

Exception in thread "main" java.lang.NumberFormatException: null
        at java.lang.Long.parseLong(Unknown Source)
        at java.lang.Long.parseLong(Unknown Source)
        at TwitchRecover.CLI.Handlers.ClipHandler.recoverClips(ClipHandler.java:165)
        at TwitchRecover.CLI.Handlers.ClipHandler.<init>(ClipHandler.java:47)
        at TwitchRecover.CLI.CLIHandler.main(CLIHandler.java:80)
        at TwitchRecover.CLI.CLI.main(CLI.java:30)

PaaaulZ commented:
Please send the url that failed

lorderheller commented:
Hi, Downloaded the newest exe file Paulz, it insta crashes every time for me. The original version starts and works normally.

PaaaulZ commented:
did you try the jar file? everything works for me

Oondanomala commented:
hi, when i run the program it complains that java is missing and exits, I downloaded it but it still fails
```
This program needs Java to run.
Please download it at http://www.java.com
```

PaaaulZ commented:
Honestly I don't know what to say, the jar works fine for me and I tried in multiple computers.

Try the latest version on my repo and see if it works. This issue is getting messy and no sign of a developer to approve pull requests so I'm moving my contributions on my fork. I cleaned the code, fixed some issues and I'll be working only on that one, just don't expect too much, I don't have a lot of time.

KaMyKaSii commented:
For you that is not working, try to open cmd and run `SET path` and look for `C:\Program Files (x86)\Common Files\Oracle\Java\javapath` or `C:\Program Files\Common Files\Oracle\Java\javapath` is among the given paths. If not, maybe [PaaaulZ](https://github.com/PaaaulZ) need to put the full java path in `run.bat` for those people who for some reason installed java but don't have the folder set in PATH.

Oondanomala commented:
It works with the new release, thank you.
Could you also enable issues on the fork so i can use that one instead of this if i have any other issue?

lorderheller commented:
@PaaaulZ the newest revision works, thank you so much for your contribution.

Kaastiron commented:
Hey I've been using the old program without issue for a while now and thought I'd try this version. 
I've been having issues using the fork, for example when I try to get a link to a vod it just crashes and I get this error message 
Please enter the link: https://www.twitch.tv/videos/1437304836
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
        at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
        at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
        at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
        at java.base/java.util.Objects.checkIndex(Objects.java:359)
        at java.base/java.util.ArrayList.get(ArrayList.java:427)
        at TwitchRecover.Core.API.VideoAPI.getSubVODFeeds(VideoAPI.java:80)
        at TwitchRecover.Core.VOD.getVODFeeds(VOD.java:113)
        at TwitchRecover.CLI.Handlers.VODHandler.retrieve(VODHandler.java:38)
        at TwitchRecover.CLI.Handlers.VODHandler.<init>(VODHandler.java:18)
        at TwitchRecover.CLI.CLIHandler.main(CLIHandler.java:24)
        at TwitchRecover.CLI.CLI.main(CLI.java:13)
        
I then tried a different VOD from a different channel (https://www.twitch.tv/videos/1435852147) and it worked. The old program  still works with all channels for me.

I also cannot recover VODS with this version as I tried to recover this https://twitchtracker.com/gmhikaru/streams/44986082012 but it only worked in the old program, in the new program it just says 
NO RESULTS FOUND.
The VOD cannot be found on Twitch servers.
Which is false as I was able to get the link (https://dgeft87wbj63p.cloudfront.net/4f4ed2686190a54b1559_gmhikaru_44986082012_1647532381/chunked/index-dvr.m3u8) from the old program.

This was all on the Latest version of your fork, I tried to download and use the older version of the fork but the exe file wouldn't open for me. I have JDK 18 installed so I'm not sure what's going wrong.
Thanks for taking your time to update this program btw.

PaaaulZ commented:
I checked the code better and the entire software is against Twitch TOS especially the sub-only bypass. StreamsCharts added some protections too breaking everything all over again.

I deleted everything to avoid problems with Twitch and I won't work on it anymore. I'm sorry for the lack of notice but I really don't want legal problems :)

This pull request is staying up just because I can't delete it, I appreciate everyone following the project and asking about the fork but I just can't work on it anymore

Oondanomala commented:
if anyone still needs this, here are some downloads, thanks to archive.org:
[Latest release archived](https://web.archive.org/web/20220405063515/https://github.com/PaaaulZ/TwitchRecover/releases/download/v1.0.1/TwitchRecover.zip)
[Latest source archive](https://web.archive.org/web/20220405063515/https://github.com/PaaaulZ/TwitchRecover/archive/refs/tags/v1.0.1.zip), afaik there was no more activity after that.
The code for this branch is still here :)
